src/transition.S

```diff
@@ -47,23 +47,42 @@ comp_exec_in:
 
     // Load params (only handle 3 params):
 loading_params:
-    cbz x4, loaded_params
-    ldr x0, [x3], #8
-    ldr x0, [x0]
-    sub x4, x4, #1
+    cbz  x4, loaded_params
+    ldrb w6, [x5], #1 // Get size of argument to load in `x6`
+                      // Shift `x5` by 1 byte since it's `char`s
+    bl   load_param_by_size
+    mov  x0, x8       // Store one argument
+    add  x3, x3, x6   // Shift array pointer by size of argument
+    sub  x4, x4, #1
 
-    cbz x4, loaded_params
-    ldr x1, [x3], #8
-    ldr x1, [x1]
-    sub x4, x4, #1
+    cbz  x4, loaded_params
+    ldrb w6, [x5], #1
+    bl   load_param_by_size
+    mov  x1, x8
+    add  x3, x3, x6
+    sub  x4, x4, #1
 
-    cbz x4, loaded_params
-    ldr x2, [x3], #8
-    ldr x2, [x2]
-    sub x4, x4, #1
+    cbz  x4, loaded_params
+    ldrb w6, [x5], #1
+    bl   load_param_by_size
+    mov  x2, x8
+    add  x3, x3, x6
+    sub  x4, x4, #1
+    b    loaded_params
+
+load_param_by_size:
+    mov  x8, #8      // Compute number of zero bytes needed for mask = 8 - `x6`
+    sub  x8, x8, x6
+    lsl  x8, x8, 3   // Multiply to get number of bits
+    mov  x7, xzr
+    sub  x7, x7, #1  // Get all ones in `x7`
+    lsr  x7, x7, x8  // Shift in needed number of zeroes
+    ldr  x8, [x3]    // Load 64 bits of argument
+    and  x8, x8, x7  // Apply mask to truncate to argument size
+    ret
```

Comment:
Note that you'll be loading off the end of the arguments array, e.g. if the last argument is smaller than 64 bits.

I think the masking could be expressed more concisely, too. Consider something like this:

```
    neg x7, x6, LSL #3    // x7 = 0 - (x6 * 8)
    ldr x8, [x3]
    lsl x8, x8, x7
    lsr x8, x8, x7
    ret
```

The shifts (`lsl`/`lsr`) only consider the bottom bits of the shift operand.

Reply:
> Note that you'll be loading off the end of the arguments array, e.g. if the last argument is smaller than 64 bits.

I did it this way to be able to load all the arguments via one `ldr`, rather than have to `if/else` various `ldrb`/`ldrw`/`ldrsw` and however many variants are there. I thought this would be better than having to do all the checks. What do you think?

> I think the masking could be expressed more concisely, too. Consider something like this:

That looks interesting; I believe you that it functionally works the same (and half expected you to give me a more optimal solution (: ), but I'll have to take some time to convince myself it does the same thing.

Reply:
> better

Well it's simpler and probably faster, but if the out-of-bounds read looks at unmapped memory, or memory outside the switcher's DDC, it would abort. This clearly doesn't happen in the tests, but I couldn't see (when reviewing) that it would never happen.

Reply:
> I'll have to take some time to convince myself it does the same thing.

In case a diagram helps:

```
Let's say that x6 = 2 (bytes), then we start with this situation:

                                                   |<--  x6 * 8 -->|
x8: ................................................aaaaaaaaaaaaaaaa

First, work out how many high bits we want to discard:

neg x7, x6, LSL #3    // x7 = 0 - (x6 * 8)

   |<---------            x7               ------->|

Then just shift left, and then back again, to discard it. (`lsr` shifts in
zeroes on the left.)


lsl x8, x8, x7
x8: aaaaaaaaaaaaaaaa000000000000000000000000000000000000000000000000

lsr x8, x8, x7
x8: 000000000000000000000000000000000000000000000000aaaaaaaaaaaaaaaa
```

Reply:
_I've left this unresolved because I think you said (in the call) that you were working on a test, or otherwise still thinking about the over-read issue._